### PR TITLE
[occm] Support for LB providers that do not implement the bulk update/create API call

### PR DESCRIPTION
**Pull request by m-bull**
**What this PR does / why we need it**:
Add support for LB providers that do not implement the bulk create/update API call.

**Which issue this PR fixes(if applicable)**:
fixes https://github.com/kubernetes/cloud-provider-openstack/issues/2271

**Special notes for reviewers**:
As this PR adds new configuration options for both Octavia ingress and Loadbalancers, additional configuration is required to test:

```
[LoadBalancer]
provider-requires-serial-api-calls=true
```
and
```bash
cat <<EOF > /etc/kubernetes/octavia-ingress-controller/config.yaml
---
kind: ConfigMap
apiVersion: v1
metadata:
  name: octavia-ingress-controller-config
  namespace: kube-system
data:
  config: |
    cluster-name: ${cluster_name}
    openstack:
      auth-url: ${auth_url}
      domain-name: ${domain-name}
      username: ${user_name}
      # user-id: ${user_id}
      password: ${password}
      project-id: ${project_id}
      region: ${region}
    octavia:
      subnet-id: ${subnet_id}
      floating-network-id: ${public_net_id}
      provider-requires-serial-api-calls: true
EOF
kubectl apply -f /etc/kubernetes/octavia-ingress-controller/config.yaml
```

**Release note**:

```release-note
[openstack-cloud-controller-manager] Adds provider-requires-serial-api-calls config option to support Octavia providers that do not implement the bulk update/create API call. Set provider-requires-serial-api-calls to true to use serial calls to Octavia to create and populate loadbalancers.
[octavia-ingress-controller] Adds provider-requires-serial-api-calls config option to support Octavia providers that do not implement the bulk update/create API call. Set provider-requires-serial-api-calls to true to use serial calls to Octavia to create and populate loadbalancers.
```

### CODE REVIEW

File: pkg/util/util.go [hunk 1]
```diff
@@ -16,6 +16,15 @@ import (
 	"k8s.io/klog/v2"
 )
 
+// CutString makes sure the string length doesn't exceed 255, which is usually the maximum string length in OpenStack.
+func CutString(original string) string {
```
Comment (dulek) [inline pkg/util/util.go:20]: If this lands on utils you got to either name this something like `CutStringTo255()` or allow passing `255` as a parameter.

File: pkg/openstack/loadbalancer.go [hunk 2]
```diff
@@ -451,6 +451,57 @@ func getListenerProtocol(protocol corev1.Protocol, svcConf *serviceConfig) liste
...
+			createOpts.VipNetworkID = svcConf.lbNetworkID
+		} else {
+			klog.V(4).Infof("network-id parameter not passed, it will be inferred from subnet-id")
+		}
+	}
+	// For external load balancer, the LoadBalancerIP is a public IP address.
+	loadBalancerIP := service.Spec.LoadBalancerIP
+	if loadBalancerIP != "" && svcConf.internal {
+		createOpts.VipAddress = loadBalancerIP
+	}
```
Comment (dulek) [inline pkg/openstack/loadbalancer.go:495]: Looks like this whole fragment is duplicated in lines 506-549. Could you extract it to another function and use it in both places? Something like `getOctaviaLoadBalancerCreateOpts()`.

**Review by dulek — CHANGES_REQUESTED** [review]
This looks nice and clean! Couple of remarks inline. Please also make sure to add the information about the new configuration options to the docs: https://github.com/kubernetes/cloud-provider-openstack/blob/dd53a6e535056322548361cbf3345d9590bcc304/docs/openstack-cloud-controller-manager/using-openstack-cloud-controller-manager.md?plain=1#L268-L282

File: pkg/util/util.go (hunk 1 above)
Comment (m-bull) [reply pkg/util/util.go]: Renamed to `CutStringTo255()`.

File: pkg/openstack/loadbalancer.go (hunk 2 above)
Comment (m-bull) [reply pkg/openstack/loadbalancer.go]: I ended up switching this around a bit to reuse the original function and just add the "fully-populated" createOpts only when the provider doesn't require the serial-style requests.

Happy to do as you suggest here if this approach isn't as good in your opinion.

Comment (dulek) [reply pkg/openstack/loadbalancer.go]: Oh no, this is nice and clean, great idea!

**Comment by m-bull** [comment]
@dulek - thanks for the reviews.

I'm leaving this as [WIP] just for a minute while I do some more testing on the system that I originally made this patch for, will update the title as soon as I can confirm that everything still works as intended with the new options set.

**Comment by m-bull** [comment]
@dulek - I made a final change to correct a log message and tested this patch successfully on the cloud that I wrote the original one for. `[WIP]` removed.

**Comment by dulek** [comment]
/lgtm

I'd still like to have @mdbooth opinion on option naming here.

**Comment by mdbooth** [comment]
> provider-requires-serial-api-calls

I think it's specific enough to not be worth bike-shedding further.

**Pull request merged**